timemachine/fe/reweighting.py

```diff
@@ -0,0 +1,289 @@
+__all__ = [
+    "one_sided_exp",
+    "construct_endpoint_reweighting_estimator",
+    "construct_mixture_reweighting_estimator",
+    "interpret_as_mixture_potential",
+]
+
+from typing import Callable, Collection
+
+from jax import numpy as np
+from jax.scipy.special import logsumexp
+
+Samples = Collection
+Params = Collection
+Array = np.ndarray
+Energies = Array
+
+BatchedReducedPotentialFxn = Callable[[Samples, Params], Energies]
+
+
+def log_mean(log_values: Array) -> float:
+    """stable log(mean(values))
+
+    log(mean(values))
+    = log(sum(values / len(values)))
+    = logsumexp(log(values) - log(len(values))
+    """
+    return logsumexp(log_values - np.log(len(log_values)))
+
+
+def estimate_log_z_ratio(log_importance_weights: Array) -> float:
+    """stable log(mean(importance_weights))"""
+    return log_mean(log_importance_weights)
+
+
+def one_sided_exp(delta_us: Array) -> float:
+    """exponential averaging
+
+    References
+    ----------
+    [1] pymbar implementation
+        https://github.com/choderalab/pymbar/blob/15f932a271343e611ed4be2d468c77b1d11cf01f/pymbar/exp.py#L54
+    """
+    # delta_us = -log_importance_weights
+    # delta_f  = -log_z_ratio
+    return -estimate_log_z_ratio(-delta_us)
+
+
+def interpret_as_mixture_potential(u_kn: Array, f_k: Array, N_k: Array) -> Array:
```

Comment:
nit: could this be relaxed to
```suggestion
def interpret_as_mixture_potential(u_kn: Array, f_k: Array, N_k: list) -> Array:
```

Reply:
(this is an extreme nit and the suggestion probably not what you intended -- just noticed we call `np.array` on the argument below, and pass in lists in the test code)

Reply:
Waffling between relaxing the signature to something more accurate like `Sequence` or Union[List, Array]` or similar vs. restricting the usages to `Array`...

Just to keep the signature simple / uniform, I'd prefer to just restrict the usages in the tests/ to `Array` (but keep the defensive array conversion in case someone accidentally passes in a list instead of an array)?

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/654/commits/59d8a6356159fa3fecfbbdd7e65db7ca14ff1132 !